packages/vite/src/node/assetSource.ts

```diff
@@ -0,0 +1,231 @@
+import type { DefaultTreeAdapterMap, Token } from 'parse5'
+
+interface HtmlAssetSource {
+  tag: string
+  attributes: string[]
+  type: 'src' | 'srcset'
+  filter?: (data: HtmlAssetSourceFilterData) => boolean
+}
+
+interface HtmlAssetSourceFilterData {
+  attribute: string
+  value: string
+  attributes: Record<string, string>
+}
+
+interface AssetAttribute {
+  attribute: Token.Attribute
+  type: 'src' | 'srcset'
+  location: Token.Location
+}
+
+export function getNodeAssetAttributes(
+  node: DefaultTreeAdapterMap['element']
+): AssetAttribute[] {
+  const assetAttrs: AssetAttribute[] = []
+
+  for (const assetSource of DEFAULT_HTML_ASSET_SOURCES) {
+    if (assetSource.tag !== node.nodeName) continue
+
+    const attributes = node.attrs.reduce((acc, attr) => {
+      acc[getAttrKey(attr)] = attr.value
+      return acc
+    }, {} as Record<string, string>)
+
+    const attrNames = Object.keys(attributes)
+
+    for (let i = 0; i < node.attrs.length; i++) {
+      const attr = node.attrs[i]
+
+      if (
+        attr.value &&
+        assetSource.attributes.includes(attrNames[i]) &&
+        (!assetSource.filter ||
+          assetSource.filter({
+            attribute: attr.name,
```

Comment:
I guess this needs to be `attrNames[i]`.

Reply:
```suggestion
            attribute: attrNames[i],
```